README.md
# Graylog cookbook

This cookbook sets up [Graylog](http://graylog.org).

Including the following support services:

- Elasticsearch
- MongoDB

# Quickstart

To quickly setup a working Graylog installation on a single node, do the following:

1. Setup application secrets

  This is required, as it would leave your Graylog installation insecure. Therefore the cookbook
  will fail with an error message if you do not set them!

  ```ruby
  # Set this to a random string, generated e.g. with "pwgen 96"
  node['graylog']['server']['server.conf']['password_secret'] = 'CHANGE ME!'

  # Generate with "echo -n yourpassword | shasum -a 256"
  node['graylog']['server']['server.conf']['root_password_sha2'] = '...'
  ```

2. Add default recipe to your run\_list

  ```json
  {
    "run_list": [
      "recipe[graylog::default]"
    ]
  }
  ```


## Supported Platforms

Currently tested on Ubuntu-14.04 LTS.

## Dependencies

- Chef `>= 0.11`
- [MongoDB cookbook](https://github.com/hipsnip-cookbooks/mongodb)
- [Apt cookbook](https://github.com/opscode-cookbooks/apt)


## Notes

Please do not expose the Graylog service directly in production. Instead, you
should use a reverse proxy (e.g. [nginx](http://nginx.org)).
This also adds the capability to use SSL to secure your logins.

Here's an example nginx site configuration you can use:j
Note: The configuration assumes you set the `rest_listen_url` accordingly:

```ruby
default['graylog']['server']['server.conf']['web_endpoint_uri'] = 'https://graylog.example.com/api/'
```

```
# Upstream to Graylog frontend
proxy_next_upstream error timeout;
upstream graylog_web_interface {
    server 127.0.0.1:9000 fail_timeout=0;
}

upstream graylog_rest_api {
    server 127.0.0.1:12900 fail_timeout=0;
}

# Redirect everything to https
server {
    listen 80;

    return 301 https://graylog.example.com$request_uri;
}

server {
    listen 443 ssl;

    # SSL certificate
    ssl_certificate /etc/nginx/certs/graylog.example.com.crt;
    ssl_certificate_key /etc/nginx/certs/graylog.example.com.key;

    root /usr/share/nginx/html;

    proxy_pass_header Date;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $http_host:$server_port;
    proxy_redirect off;
    proxy_set_header X_FORWARDED_PROTO $scheme;

    chunked_transfer_encoding off;

    location / {
        proxy_pass http://graylog_web_interface;
    }

    location /api {
        rewrite ^/api(.*)$ $1 break;
        proxy_pass http://graylog_rest_api;
    }
}
```


## Attributes

Global attribute:

```ruby
# Graylog version to use (must be available via the official repositories)
node['graylog']['version'] = '1.3'
```

### Graylog server

Attributes to configure Graylog.
The `password_secret` and `root_password_sha2` attributes NEED to be changed!

```ruby
# You MUST set a secret to secure/pepper the stored user passwords here. Use at least 64 characters.
# Generate one by using for example: pwgen -s 96
node['graylog']['server']['server.conf']['password_secret'] = 'CHANGE ME!'

# the default root user is named 'admin'
# You MUST specify a hash password for the root user (which you only need to initially set up the
# system and in case you lose connectivity to your authentication backend)
# This password cannot be changed using the API or via the web interface.
# Create one by using for example: "echo -n yourpassword | shasum -a 256"
#
# For testing purposes (only!) you can use the password "insecure" with the following hash
node['graylog']['server']['server.conf']['root_password_sha2'] = '1d92dae504a70fbcae6d3721a55d7eacaf94d3133ea5f0394b7d203d64841110'
```

This recipe disables multicast to learn about Elasticsearch. This is [recommended for production](http://support.torch.sh/help/kb/graylog2-server/configuring-and-tuning-elasticsearch-for-graylog2-v0200).

```ruby
# The default unicast host used and configured by this recipe is automatically retrieved from the Elasticsearch attributes
# (See below, node['graylog']['elasticsearch']['host'] and node['graylog']['elasticsearch']['port'])
node['graylog']['server']['server.conf']['elasticsearch_discovery_zen_ping_multicast_enabled'] = false
node['graylog']['server']['server.conf']['elasticsearch_discovery_zen_ping_unicast_hosts'] = '127.0.0.1:1234'
```

The cookbook accepts every possible configuration option supported by server.conf

```ruby
node['graylog']['server']['server.conf']['key'] = 'value'
```


### Elasticsearch

The `elasticsearch` recipe installs Elasticsearch using the official PPA repository.
You can finetune the installation here, although the defaults should "just work".

The settings below are the defaults

```ruby
# Elasticsearch version to use. Currently 0.90.x and 1.0.x versions are available
# See: http://www.elasticsearch.org/blog/apt-and-yum-repositories/
node['graylog']['elasticsearch']['version'] = '0.90'

# Assign half of the systems memory to elasticsearch heap (recommended setting)
# See: http://support.torch.sh/help/kb/graylog2-server/configuring-and-tuning-elasticsearch-for-graylog2-v0200
node['graylog']['elasticsearch']['heap_size'] = "#{(node['memory']['total'].to_i / 1024 / 2).to_i}m"
node['graylog']['elasticsearch']['cluster_name'] = 'graylog2'

# Elasticsearch ip:port to use
node['graylog']['elasticsearch']['host'] = '127.0.0.1'
node['graylog']['elasticsearch']['port'] = 9300
```


### MongoDB

The `default` recipe installs MongoDB, using the [MongoDB cookbook](https://github.com/hipsnip-cookbooks/mongodb).

As MongoDB is only used to store small amounts of data, it's usually sufficient to use a small data partition. Therefore, smallfile is enabled by default.
You can override the setting if needed, like so

```ruby
node['mongodb']['config']['smallfiles'] = false
```


## Recipes

### graylog::default

Installs and configures Elasticsearch, MongoDB, Graylog server and The Graylog web-interface.

### graylog::elasticsearch

Installs Elasticsearch from the official PPA, and configures it for Graylog use.

### graylog::server

Installs and configures Graylog server.


## Contributing

1. Fork the repository on Github
2. Create a named feature branch (i.e. `add-new-recipe`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request

## License and Authors

Author:: Chris Aumann (<me@chr4.org>)


CHANGELOG.md
Changelog for Graylog cookbook

2.0.0
-----

- Upgrade default Elasticsearch version to 2.3
- Upgrade default Graylog version to 2.0
- Fix issue that graylog-server was not restarted automatically upon config changes
- New default name `graylog` instead of `graylog2` was not adapted to make migration easier
- Upgrade Java version to openjdk-8, install `openjdk-r` PPA on trusty
- Adapt nginx example in README

1.3.3
-----

- Add support for `graylog2.gzip-filter` and `graylog2.client.accept-any-certificate` web interface attributes

1.3.2
-----

- Include `apt_repository` recipe for easier handling in `server` and `web recipes`


1.3.1
-----

- Change default spooling dir to `/var/lib/graylog-server`, according to package default
- Update other default settings to match Graylog defaults


1.3.0
-----

- Update default Graylog version to `1.3.0`

1.2.1
-----

- Use Elasticsearch `1.7` by default

1.2.0
-----

- Update default Graylog version to `1.2.0`
- Use official repositories to install `graylog-server` and `graylog-web` packages
- Rename `web_interface` recipe to `web`, to follow package naming
- Rename `graylog.conf` attributes to `server.conf` to follow package naming
- Rename `graylog-web-interface.conf` attribtues to `web.conf` to follow package naming
- Add newly introduced options to attributes

1.1.6
-----

- Update default Graylog version to `1.1.6`

1.1.5
-----

- Update default Graylog version to `1.1.5`

1.1.4
-----

- Update default Graylog version to `1.1.4`

1.1.3
-----

- Update default Graylog version to `1.1.3`

1.1.1
-----

- Update default Graylog version to `1.1.1`
- Update default elasticsearch version to `1.5`

1.0.0
-----

- Update default Graylog version to `1.0.0`
- Configuration option and config file changes according to version `1.0.0`

  Attribute names for configruation files changed:
    graylog2.conf -> graylog.conf
    graylog2-web-interface.conf -> graylog-web-interface.conf

  Node ID default URL changed (Migrate the file manually, if needed)
    /etc/graylog2-server-node-id -> /etc/graylog/server/node-id

0.92.3
------

- Update default Graylog2 version to `0.92.3`

0.92.0
------

- Update default Graylog2 version to `0.92.0`
- Update Elasticsearch `1.4`

0.91.3
------

- Update default Graylog2 version to `0.91.3`


0.91.1
------

- Update default Graylog2 version to `0.91.1`

0.91.0
------

- Update default Graylog2 version to `0.91.0`
- Update Elasticsearch to `1.3`

0.20.6
------

- Update default Graylog2 version to `0.20.6`
- Disable default secrets/passwords and display helpful error message if not set

0.20.3
------

- Initial release of graylog
- Use default Graylog2 version `0.20.3`


recipes/server.rb
#
# Cookbook Name:: graylog
# Recipe:: server
#
# Copyright (C) 2014 Chris Aumann
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

include_recipe 'graylog::apt_repository'

# Check whether password secret and root password is set
unless node['graylog']['server']['server.conf']['password_secret']
  fail <<-EOS
    Password secret is not set!
    Please set the attribute `default['graylog']['server']['server.conf']['password_secret'] = 'CHANGE ME!'`
    In your node configuration or wrapper cookbook! Use at least 64 characters.
    Generate one by using for example: "pwgen -s 96".
  EOS
end

unless node['graylog']['server']['server.conf']['root_password_sha2']
  fail <<-EOS
    Admin password is not set!
    Please set the attribute `default['graylog']['server']['server.conf']['root_password_sha2'] = '...'`
    In your node configuration or wrapper cookbook!
    Generate it with "echo -n yourpassword | shasum -a 256".

    For testing purposes only (!) you can use the following hash (password: "insecure")
    "1d92dae504a70fbcae6d3721a55d7eacaf94d3133ea5f0394b7d203d64841110"
  EOS
end

package 'graylog-server'

# Configuration
template '/etc/graylog/server/server.conf' do
  mode      00644
  source    'graylog/server.conf.erb'
  variables config: node['graylog']['server']['server.conf']
end

# Enable and start service
service 'graylog-server' do
  supports restart: true
  action [:enable, :start]

  subscribes :restart, 'template[/etc/graylog/server/server.conf]'
end
